posts/bayesian-model-comparison/index.qmd

```diff
@@ -0,0 +1,470 @@
+However if we can specify costs then decision making based on these would be the best way to work as this allows one to take into consideration the utility of the decision that will be made.
+
+In our situation, in order to do cost-sensitive analysis/decision making, all we need is to specify a cost function -- we won't need to run an optimisation on this objective function. 
+This is a function that defines the loss (or cost) we incur in making a given decision (e.g. the wrong decision).
+A typical example is whether to give more importance to a false positive or a false negative.
+For our given situation and for the sake of exposition let's assume that the random forest model takes 4 times longer to train. 
+There are three decisions we can make 
+
+1) SVM is better than RF
+2) SVM is equivalent to RF
+3) SVM is worse than RF
+
+Consider the following cost-matrix:
+
++---------------+---------------+--------------------+--------------------+
+|               | SVM is better | are equivalent     | RF is better       |
++===============+===============+====================+====================+
+| Choose SVM    | 0             | -5                 | 2                  |
+|               |               |                    |                    |
++---------------+---------------+--------------------+--------------------+
+| Choose RF     | 7             | 5                  | 0                  |
+|               |               |                    |                    |
++---------------+---------------+--------------------+--------------------+
+
+where the $(i, j)$th entry is the cost incurred in making decision $i$ when $j$ is correct.
+Here we have a $2 \times 3$ matrix as we only consider the options of selecting the SVM or the RF, no abstention or anything else.
+
+In this case we have:
+
+* cost of choosing SVM is:
+  + 0 if SVM is better, 
+  + -5 if they are equivalent as we save compute time, 
+  + 2 if RF is better as we would lose performance
+
+* cost of choosing RF is:
+  + 7 if SVM is better because we pay for computational and performance cost, 
+  + 5 if they are equivalent as we add to the compute time, 
+  + 0 if RF is indeed better
+
+
+The **expected cost** can then be obtained by multiplying the cost_matrix with the posterior probabilities
+```{python}
+cost_matrix = np.array([
+  [0., -5., 2.],
+  [7., 5., 0.],
+])
+
+probabilities = np.array([prob_svm_better, prob_rope, prob_rf_better])
+
+expected_cost = cost_matrix.dot(probabilities)
+```
+```{python}
+#| echo: false
+print(f"\nCost of deciding on SVM:", round(expected_cost[0], 4))
+print(f"\nCost of deciding on RF:", round(expected_cost[1], 4))
+```
+
+The lowest cost would determine the optimal decision.
+We see that we would incur a significant cost in choosing the SVM over the RF. 
```

Comment:
i wonder whether it would have been interesting to think up a hypothethical in which the difference is reversed, or not as pronounced when taking into account the cost, despite the probability being so skewed.

Reply:
That's certainly possible. However in this case the models are so similar in latency and memoryy etc that it's hard to justify. If RF had been neural then I think one could concoct costs so extreme that you would get what you want. 

I think the options for something more interesting are:
1) chose models for which distribution isn't so much towards the right
2) using a larger rope
3) using "graded" partitions for the cost. Something I'd mentioned to you and that I'd like to show you

Reply:
i think we are both on the same page, so i say we drop my comment and leave as is, as 1 and 2 are much hassle for little gain, and 3 is very cool but we should do in the next post.